### DISCUSSION TITLE
Why use return type TupleGenerator in `BaseModel.__iter__`?

### DISCUSSION THREAD

**Discussion by http403** [pydantic/pydantic#5282] [General]
I understand that the original implementaion of `__iter__` will statisify the defination of `TupleGenerator` since it's retrun from `__dict__`. However, when using custom root types with Sequence instead of Mapping, the return will not statisify the return type. It's not a big deal but Pylance complain to me that `return type mismatch` and `Method "__iter__" overrides class "BaseModel" in an incompatible manner` which caused me to investigate what's wrong. Will using a generic `typing.Generator[Any, None, None]` better and less misleading?

`BaseModel.__iter__`: https://github.com/pydantic/pydantic/blob/a5c63edaa1d6d238ed107e05cc8e6080f9664f24/pydantic/main.py#L556-L560
`TupleGenerator` defination: https://github.com/pydantic/pydantic/blob/a5c63edaa1d6d238ed107e05cc8e6080f9664f24/pydantic/main.py#L45

**Comment by samuelcolvin** [comment]
We won't change this in V1, dropping (technically deprecating) `__root__` in V2.

@dmontagu - another reason to deprecate `__root__` in V2.

**Reply by http403** [reply]
Thanks for the speedy response. I will look forward to v2 release.

**Comment by lucasgadams** [comment]
This still seems like a problem in V2? What is a good way to override the __iter__ method for list root classes? Or is that a bad idea? Here is a sample class:
```
class RootModelList(RootModel[list[T]]):
    root: list[T] = []

    def __iter__(self):
        return iter(self.root)

    def __getitem__(self, item: int):
        return self.root[item]

    def __len__(self):
        return len(self.root)
```

Pyright complains about the iter override not being TupleGenerator. Should we not do things like this?

**Reply by AutonomousCat** [reply]
@lucasgadams Did you figure out how to override it? I'm trying to work with dicts not tuples

**Comment by schlich** [comment]
I found it much easier to just ditch the RootModel type and make a real BaseModel.  to be honest, i'm not sure what the purpose of RootModel is other than to save a few characters on a keyword argument (characters that you lose by needing to explicitly define the base class anyways.

IMO the documentation could provide a little more motivation for this abstraction.
